jira_integration_example/utilities/alert.py

```diff
@@ -0,0 +1,250 @@
+        point.interval.end_time.nanos = int(
+            (now - point.interval.end_time.seconds) * 10**9)
+        self._client.create_time_series(self._project_name, [series])
+
+
+    def delete_custom_metric(self, metric_name):
+        """Delete the given custom metric.
+
+        Args:
+            metric_name: Name of metric to delete
+        """
+        descriptor_name = f'{self._project_name}/metricDescriptors/custom.googleapis.com/{metric_name}'
+        self._client.delete_metric_descriptor(descriptor_name)
+        print(f'Deleted metric descriptor {descriptor_name}.')
+
+
+class TestPolicyClient():
+    """Client that can create and modify alerting policies and
+    trigger/resolve incidents for testing purposes.
+
+    Attributes:
+        _project_id: The id of the project to add or modify policies to
+        _policy_client: A monitoring_v3.AlertPolicyServiceClient instance to call the alertPolicies API
+        _metric_client: A TestCustomMetricClient() instance to modify metric data
+        to trigger/resolve incidents
+        _threshold_value: Value above which a policy triggers an incident
+    """
+
+    def __init__(self, project_id, metric_client):
+        self._project_id = project_id
+        self._policy_client = monitoring_v3.AlertPolicyServiceClient()
+        self._metric_client = metric_client
+        self._threshold_value = TRIGGER_VALUE
+        self._project_name = self._policy_client.project_path(self._project_id)
+
+
+    def get_policy(self, policy_name):
+        """Get the policy with the given policy display name.
+
+        Args:
+            policy_name: the display name of the policy
+
+        Returns:
+            the AlertPolicy instance with the given policy display name
+
+        Raises:
+            PolicyNotFoundError: if the policy doesn't exist
+        """
+        try:
+            policy = list(self._policy_client.list_alert_policies(name=self._project_name, filter_=f'display_name = "{policy_name}"'))[0]
+            return policy
+        except IndexError:
+            raise PolicyNotFoundError()
+
+
+    def create_policy(self, policy_name, metric_name):
+        """Creates an alert policy with the given policy display name.
+
+        By default, a policy is made with a single condition that triggers
+        if a custom metric with the given metric name is above the threshold value.
+
+        Args:
+            policy_name: the name to identify the policy by
+            metric_name: metric to attach the policy to
+
+        Raises:
+            DuplicatePolicyError: If policy with given policy display name already exists.
+            CustomMetricNotFoundError: If the given metric doesn't exist.
+        """
+        try:
+            self.get_policy(policy_name)
+        except PolicyNotFoundError:
+            pass
+        else:
+            raise DuplicatePolicyError(f'policy with display name {policy_name} already exists.')
```

Comment:
Sorry if I'm getting ahead of myself, but I have a few implementation ideas here:

1. If the policy already exists, then log, but continue. It might mean that cleanup failed last time. Or, another test is ongoing. The problem in the second case is that the two tests might step on each others toes, and fail.
2. If the policy already exists, then crash and refuse to run the test. This solves that second problem in #1, but is annoying because you can't run the test.
3. Include a timestamp in the policy and descriptor names, so each one will be unique. This makes things a bit more complex, because there is the risk that cleanup with always fail, and over time, you will create an infinite number of metric/policies without ever cleaning them up. The solution here is, at cleanup, call list_policies() and list_metrics(), then clean up everything older than, say, a few hours that belongs to the test. The advantage here is that if two tests are running at the same time, those two tests will be completely independent of one another (for example, if two people trigger the CI at the same time)